In [1]:
from google.colab import drive 
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
import scipy.stats as stats
from scipy.stats import ttest_1samp
from statsmodels.stats.power import tt_ind_solve_power

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
data=pd.read_excel('/content/drive/MyDrive/BMI.xls')
data

Name  Age  Gender  Height  Weight   BMI Value   BMI Status
0  Shivani    21      F     164       60       22.3       Normal
1   Gaurav    22      M     180       83       25.6   Overweight
2    Hamza    25      M     176       65       21.0       Normal
3   Preeti    26      F     160       62       24.2       Normal
4   Aniket    30      M     182       90       27.2   Overweight
5    Suraj    27      M     170       45       15.6  Underweight
6    Nidhi    18      F     158       72       28.8   Overweight
7    Kayum    20      M     170       55       19.0  Underweight

### Hypothesis Testing or Significance Testing

---

Hypothesis testing is a statistical method that is used in making statistical decisions using experimental data.  Hypothesis Testing is basically an assumption that we make about the population parameter.

There are many different statistical significance, or hypothesis, tests. They all follow the same basic principle. The appropriate test for a given situation depends on the nature of the data being analysed.

***Null hypothesis:***

All statistical significance tests start with a null hypothesis. A statistical significance test measures the strength of evidence that the data sample supplies for or against some proposition of interest.

This proposition is known as a 'null hypothesis', since it usually relates to there being 'no difference' between groups' or 'no effect' of a treatment.

For example:

1.CMV infected babies have the same average birthweight as non-infected babies.

2.ß-thalassaemia does not have any effect on ferritin level.

3.Recombinant human erythropoietin has no effect on the haemoglobin levels in premature infants.

Even if our hypothesis is not to do with 'no difference' it is still convention that the hypothesis being tested is known as the ***null hypothesis***.

---
---

1) check if population mean value of BMI is significantly different than  23 . Use 0.05 as α (los).

Null Hypothesis will be

Ho : mu = 23 

Alternate Hypothesis will be

H1 : mu != 23


### **`One Sample t-test`**

This test is appropriate when a continuous variable is recorded on a single sample of individuals or items and the mean value is to be tested against some hypothesised population mean

Here we use t test as we don't know the population variance and the sample size is less than 30.

In [4]:
def hypothesisTesting(df,column,H0,H1,alpha,mean,tailTest):
    ts, pv = ttest_1samp(df[column].values, mean)
    t2pv = pv
    t1pv = pv*2
    if tailTest == 1:
        if t1pv < alpha:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
    else:
        if t2pv < alpha/2:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
 
hypothesisTesting(data,'BMI Value','mean = 23','mean != 23',0.05,23,2)

Null Hypothesis: Not Rejected
Conclusion: mean = 23



From above result we can say that population mean of BMI is equal to 23.

---
---

2) check if population mean value of Height is greater than 170 cm . Use 0.05 as α (los).

Null Hypothesis will be

Ho : mu > 170

Alternate Hypothesis will be

H1 : mu <= 170

Here we use t test as we know the poplation variance and the sample size is less than 30.

In [5]:
def hypothesisTesting(df,column,H0,H1,alpha,mean,tailTest):
    ts, pv = ttest_1samp(df[column].values, mean)
    t2pv = pv
    t1pv = pv*2
    if tailTest == 1:
        if t1pv < alpha:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
    else:
        if t2pv < alpha/2:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
 
hypothesisTesting(data,'Height','mean > 170','mean <= 170',0.05,23,1)

Null Hypothesis: Rejected
Conclusion: mean <= 170


From above result we can say that population mean of Height is greather or  equal to 170 cm.








# AB Testing

Hypothesis testing is a broader term that encompasses a/b testing. Typically a hypothesis test will have a null hypothesis which could be tested against two or more samples for significance.

A/B Testing goes a step further and tries to calculate implied impact from the experiment testing two randomized different variants. Randomization helps t o ensure bias is minimized but this isn't always easy to accomplish.

Example :

## ***Two Sample t test***




First, we’ll create two arrays to hold the measurements of each group of 20 plants:

In [6]:
group1 = np.array([14, 15, 15, 16, 13, 8, 14, 17, 16, 14, 19, 20, 21, 15, 15, 16, 16, 13, 14, 12])
group2 = np.array([15, 17, 14, 17, 14, 8, 12, 19, 19, 14, 17, 22, 24, 16, 13, 16, 13, 18, 15, 13])

Before we perform the test, we need to decide if we’ll assume the two populations have equal variances or not. As a rule of thumb, we can assume the populations have equal variances if the ratio of the larger sample variance to the smaller sample variance is less than 4:1. 

In [8]:
#find variance for each group
print(np.var(group1), np.var(group2))


7.727500000000001 12.260000000000002


The ratio of the larger sample variance to the smaller sample variance is 12.26 / 7.73 = 1.586, which is less than 4. This means we can assume that the population variances are equal.

Thus, we can proceed to perform the two sample t-test with equal variances:

The two hypotheses for this particular two sample t-test are as follows:

H0: µ1 = µ2 (the two population means are equal)

HA: µ1 ≠µ2 (the two population means are not equal)

In [9]:
#perform two sample t-test with equal variances
stats.ttest_ind(a=group1, b=group2, equal_var=True)

Ttest_indResult(statistic=-0.6337397070250238, pvalue=0.5300471010405257)

The t test statistic is -0.6337 and the corresponding two-sided p-value is 0.53005.

Because the p-value of our test (0.53005) is greater than alpha = 0.05, we fail to reject the null hypothesis of the test.

We do not have sufficient evidence to say that the mean height of plants between the two populations is different.

# Type I error

A type 1 error is also known as a false positive and occurs when a researcher incorrectly rejects a true null hypothesis. This means that your report that your findings are significant when in fact they have occurred by chance.

The probability of making a type I error is represented by your alpha level (α), which is the p-value below which you reject the null hypothesis. A p-value of 0.05 indicates that you are willing to accept a 5% chance that you are wrong when you reject the null hypothesis.

You can reduce your risk of committing a type I error by using a lower value for p. For example, a p-value of 0.01 would mean there is a 1% chance of committing a Type I error.

However, using a lower value for alpha means that you will be less likely to detect a true difference if one really exists (thus risking a type II error).

Example :

 In above eg ,
 the p value is 0.53005 which is considered as type I error.

## Type II error

A type II error is also known as a false negative and occurs when a researcher fails to reject a null hypothesis which is really false. Here a researcher concludes there is not a significant effect, when actually there really is.

The probability of making a type II error is called Beta (β), and this is related to the power of the statistical test (power = 1- β). You can decrease your risk of committing a type II error by ensuring your test has enough power.

You can do this by ensuring your sample size is large enough to detect a practical difference when one truly exists.


Example :

Taking example from two sample t test 

In [10]:
effect_size=abs((group1.mean()-group2.mean())/(group1.std()-group2.std()))
sample_size=20
alpha=0.05
ratio=1.0

statistical_power = tt_ind_solve_power(effect_size=effect_size, nobs1=sample_size, alpha=alpha, ratio=1.0, alternative='two-sided')
print(statistical_power)

0.7927312603881692


In [11]:
# type 2 error 
type_2_error=1-statistical_power
type_2_error


0.2072687396118308

The  probability of type 2 error is 0.2073